src/registry/mod.rs

```diff
@@ -49,77 +56,169 @@ pub fn write_lock<'a>() -> RwLockWriteGuard<'a, RegMap> {
-    trace!("> set: {} to {:?}", key, &value);
-    write_lock().insert(key, value);
+/// Get Json data from the registry, evaluating if a property was found.
+pub fn get_json<K>(name: &K) -> RegistryResult<(AccessFlags, Arc<Json>)>
+where String: Borrow<K>, K: Hash + Eq + Display {
+    get_data(name).map(|val| val.resolve())
+}
+
+/// Set a value to the given registry field.
+pub fn set_field(key: String, value: RegistryField)
+                 -> RegistryResult<Option<RegistryField>> {
+    let mut reg = write_lock();
+    match reg.entry(key) {
+        Entry::Occupied(mut entry) => {
+            let first_type = entry.get().get_type();
+            if first_type.can_set_from(value.get_type()) {
+                Ok(Some(entry.insert(value)))
+            }
+            else {
+                Err(RegistryError::WrongKeyType)
+            }
+        },
+        Entry::Vacant(vacancy) => {
+            vacancy.insert(value);
+            Ok(None)
+        }
+    }
 }
 
 /// Set a command to the given Command
 #[allow(dead_code)]
-pub fn set_command(key: String, command: CommandFn) {
-    trace!("< set_command: {}", key);
-    set(key, RegistryValue::Command(command))
+pub fn set_command(key: String, command: CommandFn)
+                   -> RegistryResult<Option<CommandFn>> {
+    set_field(key, RegistryField::Command(command))
+        .map(|maybe_field|
+              maybe_field.and_then(|field|
+                                    field.get_command()))
 }
 
 /// Set a value to the given JSON value.
-pub fn set_json(key: String, flags: AccessFlags, json: Json) {
-    trace!("< set_json: {}", key);
-    set(key, RegistryValue::new_json(flags, json))
+#[allow(dead_code)]
+pub fn set_json(key: String, flags: AccessFlags, json: Json)
+                -> RegistryResult<Option<(AccessFlags, Arc<Json>)>> {
+    let func: SetFn;
+    {
+        let mut reg = write_lock();
+        match reg.entry(key) {
+            Entry::Vacant(vacancy) => {
+                vacancy.insert(RegistryField::Object {
+                    flags: flags, data: Arc::new(json)
+                });
+                return Ok(None);
+            },
+            Entry::Occupied(mut entry) => {
+                let first_type = entry.get().get_type();
+                if first_type == FieldType::Object {
+                    return Ok(entry.insert(RegistryField::Object {
+                        flags: flags, data: Arc::new(json)
+                    }).as_object());
+                }
+                else if first_type == FieldType::Property {
+                    match entry.get().clone().as_property_set() {
+                        Some(fun) => func = fun,
+                        None => return Err(RegistryError::InvalidOperation)
+                    }
+                }
+                else {
+                    return Err(RegistryError::WrongKeyType);
+                }
+            }
+        }
+    }
+    func(json);
```

Comment:
Why do you set the func on L165 and then call it down here, instead of simply calling func on L165 and returning `Ok(None)`?